File: packages/redis/src/applyPresence.ts
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import type { Redis } from "./client.js";
import type { GameState } from "@anagrabble/protocol";

const SCRIPT_PATH = fileURLToPath(new URL("./scripts/apply_presence.lua", import.meta.url));
const SCRIPT = readFileSync(SCRIPT_PATH, "utf8");

export interface ApplyPresenceArgs {
  stateKey: string;
  playerId: string;
  lastSeenAt: number;
}

export type ApplyPresenceError = "GameNotFound";
export type ApplyPresenceResult = { state: GameState } | { error: ApplyPresenceError };

/** Atomic per-player `lastSeenAt` write — see
 * packages/redis/src/scripts/apply_presence.lua for the actual logic and
 * docs/decisions.md "Player presence: connected/disconnected tracking"
 * for why this needs to be atomic (a plain GET/SET could clobber a
 * concurrent gameplay mutation to the same state blob). */
export async function applyPresence(
  redis: Redis,
  args: ApplyPresenceArgs,
): Promise<ApplyPresenceResult> {
  const raw = (await redis.eval(SCRIPT, {
    keys: [args.stateKey],
    arguments: [args.playerId, String(args.lastSeenAt)],
  })) as string;

  const parsed = JSON.parse(raw) as GameState | { error: ApplyPresenceError };
  return "error" in parsed ? { error: parsed.error } : { state: parsed };
}

File: packages/redis/src/scripts/apply_presence.lua
-- Atomic presence write: patches one player's lastSeenAt without clobbering
-- whatever a concurrent mutation (TurnTile, SubmitWord) may be doing to the
-- same state blob at the same moment. See docs/decisions.md "Player
-- presence: connected/disconnected tracking" — "reachable" is always
-- derived from this timestamp at read time (apps/server/src/lobby.ts's
-- isReachable, mirrored in apply_turn_tile.lua), never tracked via a
-- scheduled timer, so this script only ever needs to record a timestamp,
-- never schedule or cancel anything. Deliberately has no commandId
-- idempotency set (unlike apply_turn_tile.lua/apply_submit_word.lua) —
-- writing the same lastSeenAt twice is harmless, there's nothing here that
-- would double-apply.
--
-- KEYS[1] state
-- ARGV[1] playerId, ARGV[2] lastSeenAt to write (ms) — callers pass `now`
-- for a heartbeat, or an already-expired value to mark a graceful close
-- stale immediately rather than waiting out the next heartbeat gap.
--
-- Returns the resulting GameState JSON, or {"error": "GameNotFound"}.
-- No-ops (returns state unchanged) if the player isn't found — a
-- late-arriving heartbeat racing the player's own removal isn't an error.

local stateRaw = redis.call('GET', KEYS[1])
if not stateRaw then
  return cjson.encode({ error = 'GameNotFound' })
end

local state = cjson.decode(stateRaw)
local playerId = ARGV[1]
local lastSeenAt = tonumber(ARGV[2])

for _, player in ipairs(state.players) do
  if player.id == playerId then
    player.lastSeenAt = lastSeenAt
    break
  end
end

local encoded = cjson.encode(state)
-- lua-cjson can't tell an emptied array from an empty object. Unlike
-- apply_turn_tile.lua (which only ever runs once `pool` already has a tile
-- in it), this script can fire before the first tile is ever turned, so
-- `pool` needs the same round-trip patch as `words` — see that script's
-- comment and packages/redis test coverage for the underlying regression.
encoded = string.gsub(encoded, '"words":{}', '"words":[]')
encoded = string.gsub(encoded, '"pool":{}', '"pool":[]')

redis.call('SET', KEYS[1], encoded)
return encoded
